File: spark/src/main/java/io/hops/examples/spark/kafka/StructuredStreamingKafka.java
package io.hops.examples.spark.kafka;

import com.google.common.base.Strings;
import com.twitter.bijection.Injection;
import io.hops.util.CredentialsNotFoundException;
import io.hops.util.HopsProducer;
import io.hops.util.HopsUtil;
import io.hops.util.SchemaNotFoundException;
import io.hops.util.spark.SparkProducer;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.apache.avro.generic.GenericRecord;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.streaming.DataStreamReader;
import org.apache.spark.sql.streaming.StreamingQuery;
import org.apache.spark.sql.streaming.StreamingQueryException;
import org.apache.spark.sql.streaming.Trigger;

/**
 *
 * <p>
 */
public class StructuredStreamingKafka {

  private static final Map<String, Injection<GenericRecord, byte[]>> RECORD_INJECTIONS = HopsUtil.getRecordInjections();
  private static final Logger LOG = Logger.getLogger(StructuredStreamingKafka.class.getName());

  public static void main(String[] args) throws StreamingQueryException, InterruptedException {
    final String type = args[0];
    //Producer
    if (!Strings.isNullOrEmpty(type) && type.equalsIgnoreCase("producer")) {
      Set<String> topicsSet = new HashSet<>(HopsUtil.getTopics());
      SparkConf sparkConf = new SparkConf().setAppName(HopsUtil.getJobName());
      JavaSparkContext jsc = new JavaSparkContext(sparkConf);
      final List<HopsProducer> sparkProducers = new ArrayList<>();
      final DateFormat sdf = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss:SSS");
      final List<String> messages = new ArrayList();
      final List<String> priorities = new ArrayList();
      final List<String> loggers = new ArrayList();

      /**
       * ********************************* Setup dummy test data ***********************************
       */
      messages.add("Container container_e01_1494850115055_0016_01_000002 succeeded");
      messages.add("Container container_e01_1494850115251_0015_01_000002 succeeded");
      messages.add(
          "rollingMonitorInterval is set as -1. The log rolling mornitoring interval is disabled. The logs will be aggregated after this application is finished.");
      messages.add(
          "rollingMonitorInterval is set as -1. The log rolling mornitoring interval is disabled. The logs will be aggregated after this application is finished.");
      messages.add(
          "Sending out 2 container statuses: [ContainerStatus: [ContainerId: container_e01_1494850115055_0016_01_000001, State: RUNNING, Diagnostics: , ExitStatus: -1000, ], ContainerStatus: [ContainerId: container_e01_1494850115055_0016_01_000002, State: RUNNING, Diagnostics: , ExitStatus: -1000, ]]");
      messages.add("Node's health-status : true");
      messages.add("Cannot create writer for app application_1494433225517_0008. Skip log upload this time.");
      priorities.add("INFO");
      priorities.add("INFO");
      priorities.add("WARN");
      priorities.add("DEBUG");
      priorities.add("DEBUG");
      priorities.add("DEBUG");
      priorities.add("ERROR");
      loggers.add("org.apache.hadoop.yarn.server.nodemanager.containermanager.monitor.ContainersMonitorImpl");
      loggers.add("org.apache.hadoop.yarn.server.nodemanager.containermanager.monitor.ContainersMonitorImpl");
      loggers.add("org.apache.hadoop.yarn.server.nodemanager.containermanager.logaggregation.AppLogAggregatorImpl");
      loggers.add("org.apache.hadoop.yarn.server.nodemanager.NodeStatusUpdaterImpl");
      loggers.add("org.apache.hadoop.yarn.server.nodemanager.NodeStatusUpdaterImpl");
      loggers.add("org.apache.hadoop.yarn.server.nodemanager.NodeStatusUpdaterImpl");
      loggers.add("org.apache.hadoop.yarn.server.nodemanager.containermanager.logaggregation.AppLogAggregatorImpl");
      //End setup dummy data

      //Get a broker for the producer
      LOG.log(Level.INFO, "Producing to:{0}", HopsUtil.getBrokerEndpointsList().get(0));
      Properties props = new Properties();
      props.put("bootstrap.servers", HopsUtil.getBrokerEndpointsList().get(0));
      for (final String topic : topicsSet) {
        new Thread() {
          @Override
          public void run() {
            try {
              SparkProducer sparkProducer = HopsUtil.getSparkProducer(topic, props);
              sparkProducers.add(sparkProducer);
              Map<String, String> message = new HashMap<>();
              int i = 0;
              //Produce Kafka messages to topic
              while (true) {
                message.put("message", messages.get(i % messages.size()));
                message.put("priority", priorities.get(i % priorities.size()));
                message.put("logger", loggers.get(i % loggers.size()));
                Date date = new Date();
                message.put("timestamp", sdf.format(date));
                sparkProducer.produce(message);
                Thread.sleep(100);
                i++;
              }
            } catch (SchemaNotFoundException | CredentialsNotFoundException | InterruptedException ex) {
              LOG.log(Level.SEVERE, ex.getMessage(), ex);
            }
          }
        }.start();
      }//Keep application running
      HopsUtil.shutdownGracefully(jsc);
      for (HopsProducer hopsProducer : sparkProducers) {
        hopsProducer.close();
      }
      //Consumer
    } else {
      // Create DataSet representing the stream of input lines from kafka
      DataStreamReader dsr = HopsUtil.getSparkConsumer().getKafkaDataStreamReader();
      Dataset<Row> lines = dsr.load();

      // Generate running word count
      Dataset<LogEntry> logEntries = lines
          .map(new MapFunction<Row, LogEntry>() {
            @Override
            public LogEntry call(Row record) throws Exception {
              GenericRecord genericRecord = RECORD_INJECTIONS.entrySet().iterator().next().getValue().invert(record.
                  getAs("value")).get();
              LogEntry logEntry = new LogEntry(genericRecord.get("timestamp").toString(),
                  genericRecord.get("priority").toString(),
                  genericRecord.get("logger").toString(),
                  genericRecord.get("message").toString());
              return logEntry;
            }
          }, Encoders.bean(LogEntry.class));

      Dataset<String> logEntriesRaw = lines
          .map(new MapFunction<Row, String>() {
            @Override
            public String call(Row record) throws Exception {
              GenericRecord genericRecord = RECORD_INJECTIONS.entrySet().iterator().next().getValue().invert(record.
                  getAs("value")).get();

              return genericRecord.toString();
            }
          }, Encoders.STRING());

      // Start running the query that prints the running counts to the console
      StreamingQuery queryFile = logEntries.writeStream()
          .format("parquet")
          .option("path", "/Projects/" + HopsUtil.getProjectName() + "/Resources/data-parquet-" + HopsUtil.getAppId())
          .option("checkpointLocation", "/Projects/" + HopsUtil.getProjectName() + "/Resources/checkpoint-parquet-"
              + HopsUtil.getAppId())
          .trigger(Trigger.ProcessingTime(10000))
          .start();

      StreamingQuery queryFile2 = logEntriesRaw.writeStream()
          .format("text")
          .option("path", "/Projects/" + HopsUtil.getProjectName() + "/Resources/data-text-" + HopsUtil.getAppId())
          .option("checkpointLocation", "/Projects/" + HopsUtil.getProjectName() + "/Resources/checkpoint-text-"
              + HopsUtil.getAppId())
          .trigger(Trigger.ProcessingTime(10000))
          .start();

      HopsUtil.shutdownGracefully(queryFile);
    }
  }
}

File: spark/src/main/scala/io/hops/examples/spark/kafka/StructuredStreamingKafka.scala
package io.hops.examples.spark.kafka

import io.hops.util.Hops
import org.apache.spark.sql.avro._
import org.apache.spark.sql._
import java.text.SimpleDateFormat
import java.util.Date
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

import org.apache.log4j.{Level, LogManager}
import org.apache.spark.SparkConf


/**
  * Example that uses HopsUtil and Spark-Kafka structured streaming with Avro.
  * Runs a producer and a consumer, depending on the user-provided argument.
  *
  */
object StructuredStreamingKafka {

  def main(args: Array[String]): Unit = {
    // Setup logging
    val log = LogManager.getLogger(StructuredStreamingKafka.getClass)
    log.setLevel(Level.INFO)
    log.info(s"Starting Structured Streaming Kafka with Avro Job")

    if (args.length != 2) {
      log.error("Program expects at two arguments, type<producer|consumer> and a topic name")
      System.exit(1)
    } else if (args(0) != "producer" && args(0) != "consumer") {
      log.error("Valid types are <producer|consumer>")
      System.exit(1)
    }

    // Setup Spark
    var sparkConf: SparkConf = new SparkConf()
    val spark = SparkSession.builder().appName(Hops.getJobName).config(sparkConf).enableHiveSupport().getOrCreate()
    val sc = spark.sparkContext
    val kafkaClientType = args(0)
    val topic = args(1)

    if (args(0) == "producer") {
      import spark.implicits._
      //First prepare some random input data
      val data = Seq(
        Row(new SimpleDateFormat("yyyy/MM/dd HH:mm:ss:SSS").format(new Date), "INFO", "org.apache.hadoop.yarn.server.nodemanager.containermanager.monitor.ContainersMonitorImpl", "Container container_e01_1494850115055_0016_01_000002 succeeded"),
        Row(new SimpleDateFormat("yyyy/MM/dd HH:mm:ss:SSS").format(new Date), "DEBUG", "org.apache.hadoop.yarn.server.nodemanager.containermanager.logaggregation.AppLogAggregatorImpl", "Cannot create writer for app application_1494433225517_0008. Skip log upload this time."),
        Row(new SimpleDateFormat("yyyy/MM/dd HH:mm:ss:SSS").format(new Date), "WARN", "org.apache.hadoop.yarn.server.nodemanager.NodeStatusUpdaterImpl", "\"Sending out 2 container statuses: \"\n          + \"[ContainerStatus: [ContainerId: container_e01_1494850115055_0016_01_000001, State: RUNNING, \"\n          + \"Diagnostics: , ExitStatus: -1000, ], \"\n          + \"ContainerStatus: [ContainerId: container_e01_1494850115055_0016_01_000002, \"\n+ \"State: RUNNING, Diagnostics: , ExitStatus: -1000, ]]\""),
        Row(new SimpleDateFormat("yyyy/MM/dd HH:mm:ss:SSS").format(new Date), "INFO", "org.apache.hadoop.yarn.server.nodemanager.containermanager.monitor.ContainersMonitorImpl", "Container container_e01_1494850115055_0016_01_000002 succeeded")
      )

      //Schema corresponds to the schema in the Hops Kafka Schema Registry
      val schema = StructType(
        Array(
          StructField("timestamp", StringType, nullable = false),
          StructField("priority", StringType, nullable = false),
          StructField("logger", StringType, nullable = false),
          StructField("message", StringType, nullable = false)
        )
      )

      val df = spark.createDataFrame(
        spark.sparkContext.parallelize(data),
        schema
      )

      //Send df with input data to the Kafka topic in Avro format
      df.select(to_avro(struct(df.columns.map(column): _*)).alias("value")).
        write.format("kafka").
        option("kafka.bootstrap.servers", Hops.getBrokerEndpoints).
        option("kafka.security.protocol", "SSL").
        option("kafka.ssl.truststore.location", Hops.getTrustStore).
        option("kafka.ssl.truststore.password", Hops.getKeystorePwd).
        option("kafka.ssl.keystore.location", Hops.getKeyStore).
        option("kafka.ssl.keystore.password", Hops.getKeystorePwd).
        option("kafka.ssl.key.password", Hops.getKeystorePwd).
        option("kafka.ssl.endpoint.identification.algorithm", "").
        option("topic", topic).save()


    } else {
      val dfFromKafka = spark.readStream.format("kafka").
        option("kafka.bootstrap.servers", Hops.getBrokerEndpoints).
        option("subscribe", topic).
        option("startingOffsets", "earliest").
        option("kafka.security.protocol", "SSL").
        option("kafka.ssl.truststore.location", Hops.getTrustStore).
        option("kafka.ssl.truststore.password", Hops.getKeystorePwd).
        option("kafka.ssl.keystore.location", Hops.getKeyStore).
        option("kafka.ssl.keystore.password", Hops.getKeystorePwd).
        option("kafka.ssl.key.password", Hops.getKeystorePwd).
        option("kafka.ssl.endpoint.identification.algorithm", "").
        load()

      val logsDF = dfFromKafka.select(from_avro(col("value"), Hops.getSchema(topic)) as 'logs)

      logsDF.writeStream.format("parquet").
        option("path", "/Projects/" + Hops.getProjectName + "/Resources/" + topic + "-parquet").
        option("checkpointLocation", "/Projects/" + Hops.getProjectName + "/Resources/" + topic + "-checkpoint").
        start().awaitTermination()

    }

    log.info("Shutting down spark job")
    spark.close
  }


}
